README.md
This library offers the ability to unzip files in-memory.

Implementation notes

- resilience to zip bombs through configuring decompression limit
- ZIP archives read from the file system are memory mapped
- there is currently not support for zip64
- parsing is zero-copy


CHANGELOG.md
# 0.1.1

 - add docs to some methods
 - allow configuration of decompression limit

# 0.1.0

Initial release.

src/lib.rs
/*!
 * Zip file parsing library
 *
 * ```no_run
 * # use zip_rs::ZipArchive;
 * let mut archive = ZipArchive::from_path("./foo.zip").unwrap();
 *
 * for file in archive.files() {
 *   // ...
 * }
 * ```
 */

#![deny(missing_debug_implementations)]

use std::{
    borrow::Cow,
    ffi::OsStr,
    fs::File,
    io::{Read, Write},
    ops::Deref,
    path::Path,
};

#[cfg(target_family = "unix")]
use std::os::unix::ffi::OsStrExt;

#[cfg(target_family = "windows")]
use std::os::unix::ffi::OsStrExt;

pub use common::*;
pub use error::ZipParseError;
use flate2::read::DeflateDecoder;
use parse::Parser;

mod common;
mod error;
mod parse;

const KB: usize = 1024;
const MB: usize = 1024 * KB;
const GB: usize = 1024 * MB;

/// An entire ZIP archive file
#[derive(Debug)]
pub struct ZipArchive<'a, B: Deref<Target = [u8]>> {
    pub central_directory: CentralDirectory<'a>,
    parser: Parser<B>,
}

impl<'a> ZipArchive<'a, memmap::Mmap> {
    /// Parse a [`ZipArchive`] from a file path. The contents are memory mapped.
    pub fn from_path(path: impl AsRef<Path>) -> Result<Self, ZipParseError> {
        let file = File::open(path)?;
        let buffer = unsafe { memmap::Mmap::map(&file) }?;

        Self::from_buffer(buffer)
    }
}

impl<'a, B: Deref<Target = [u8]>> ZipArchive<'a, B> {
    /// Parse a [`ZipArchive`] from an existing buffer in memory
    pub fn from_buffer(buffer: B) -> Result<Self, ZipParseError> {
        let mut parser = Parser::new(buffer);

        let central_directory = parser.parse_central_directory()?;

        Ok(ZipArchive {
            central_directory,
            parser,
        })
    }

    /// Lazily iterate over the files in this archive
    pub fn files<'b>(
        &'b mut self,
    ) -> impl Iterator<Item = Result<CompressedZipFile<'a>, ZipParseError>> + 'b {
        // todo: can we avoid this clone by wrapping self.parser in RefCell?
        let files = self.central_directory.files.clone();

        files
            .into_iter()
            .map(move |file_header| self.parser.read_file(&file_header))
    }
}

#[derive(Debug, Clone)]
pub struct CentralDirectoryFileHeader<'a> {
    pub os: Os,
    pub metadata: Metadata<'a>,
    pub disk_num_start: u16,
    pub internal_attributes: InternalAttributes,
    pub external_attributes: ExternalAttributes,
    pub zip_specification_version: u8,
    pub local_header_offset: u32,
    pub comment: &'a [u8],
}

#[derive(Debug)]
pub struct EndCentralDirectory {
    pub disk_num: u16,
    pub disk_central_dir_num: u16,
    pub disk_entires: u16,
    pub total_entires: u16,
    pub central_dir_size: u32,
    pub central_dir_offset: u32,
}

#[derive(Debug, Clone)]
pub struct Metadata<'a> {
    pub version_needed: u16,
    pub compression_method: CompressionMethod,
    pub date_time_modified: DateTimeModified,
    pub flags: ZipFlags,
    pub name: &'a [u8],
    pub extra_field: &'a [u8],
    pub compressed_size: u64,
    pub uncompressed_size: u64,
    pub crc: u32,
}

/// A single compressed ZIP file
#[derive(Debug)]
pub struct CompressedZipFile<'a> {
    pub metadata: Metadata<'a>,
    contents: &'a [u8],
}

impl<'a> CompressedZipFile<'a> {
    pub fn compressed_contents(&self) -> &[u8] {
        self.contents
    }

    /// Efficiently writes decompressed contents to sink without loading full
    /// decompressed contents into memory
    ///
    /// `limit` controls the max uncompressed file size that will be accepted. A
    /// `limit` of `None` implies no limit. Note that setting too high of a limit
    /// can make decoders susceptible to DoS through ZIP bombs or other means.
    pub fn write_with_limit(
        &self,
        w: &mut dyn Write,
        limit: Option<usize>,
    ) -> Result<(), ZipParseError> {
        if Some(self.metadata.uncompressed_size as usize) >= limit {
            return Err(ZipParseError::FileTooLarge(self.metadata.uncompressed_size));
        }

        match self.metadata.compression_method.name() {
            CompressionMethodName::None => {
                w.write_all(self.contents)?;
            }
            CompressionMethodName::Deflate => {
                let mut decoder = DeflateDecoder::new(self.contents);

                let amt_read = std::io::copy(&mut decoder, w)?;

                if amt_read != self.metadata.uncompressed_size {
                    return Err(ZipParseError::Generic("failed to write full buffer"));
                }
            }
            method => todo!("unimplemented compression method {:?}", method),
        }

        Ok(())
    }

    /// Efficiently writes decompressed contents to sink without loading full
    /// decompressed contents into memory.
    ///
    /// This method uses the default limit of 8 gigabytes. See
    /// [CompressedZipFile::write_with_limit] to configure this limit.
    pub fn write(&self, w: &mut dyn Write) -> Result<(), ZipParseError> {
        self.write_with_limit(w, Some(8 * GB))
    }

    /// Decompress full contents into memory
    ///
    /// `limit` controls the max uncompressed file size that will be accepted. A
    /// `limit` of `None` implies no limit. Note that setting too high of a limit
    /// can make decoders susceptible to DoS through ZIP bombs or other means.
    pub fn decompressed_contents_with_limit(
        &self,
        limit: Option<usize>,
    ) -> Result<Cow<[u8]>, ZipParseError> {
        if Some(self.metadata.uncompressed_size as usize) >= limit {
            return Err(ZipParseError::FileTooLarge(self.metadata.uncompressed_size));
        }

        match self.metadata.compression_method.name() {
            CompressionMethodName::None => return Ok(Cow::Borrowed(self.contents)),
            CompressionMethodName::Deflate => {
                let mut out = vec![0; self.metadata.uncompressed_size as usize];

                DeflateDecoder::new(self.contents).read_exact(&mut out)?;

                Ok(Cow::Owned(out))
            }
            method => todo!("unimplemented compression method {:?}", method),
        }
    }

    /// Decompress full contents into memory
    ///
    /// This method uses the default limit of 8 gigabytes. See
    /// [CompressedZipFile::decompressed_contents_with_limit] to configure this
    /// limit.
    pub fn decompressed_contents(&self) -> Result<Cow<[u8]>, ZipParseError> {
        self.decompressed_contents_with_limit(Some(8 * GB))
    }

    /// This file's `Path` inside the ZIP archive.
    ///
    /// Note that this path may reference file paths outside the archive through
    /// the use of absolute paths or the parent directory (`..`). The full file path
    /// should not be used when interacting with the host file system if the ZIP
    /// file is untrusted.
    pub fn file_path(&self) -> &Path {
        &Path::new(OsStr::from_bytes(self.metadata.name))
    }

    /// The raw bytes of this file's path inside the ZIP archive.
    ///
    /// Note that this path may reference file paths outside the archive through
    /// the use of absolute paths or the parent directory (`..`). The full file path
    /// should not be used when interacting with the host file system if the ZIP
    /// file is untrusted.
    pub fn file_path_bytes(&self) -> &'a [u8] {
        self.metadata.name
    }

    /// The algorithm used to compress this file.
    ///
    /// This is typically [`CompressionMethodName::None`] or
    /// [`CompressionMethodName::Deflate`].
    pub fn compression_method(&self) -> CompressionMethod {
        self.metadata.compression_method
    }
}

#[derive(Debug)]
pub struct CentralDirectory<'a> {
    pub files: Vec<CentralDirectoryFileHeader<'a>>,
    pub end: EndCentralDirectory,
}

#[cfg(test)]
mod test {
    use crate::{ZipArchive, MB};

    #[test]
    #[should_panic]
    fn zip_bomb() {
        let mut bomb = ZipArchive::from_path("files/bomb.zip").unwrap();

        for file in bomb.files() {
            let file = file.unwrap();

            file.decompressed_contents_with_limit(Some(20 * MB))
                .unwrap();
        }
    }
}


src/common.rs
pub const LOCAL_FILE_SIGNATURE: [u8; 4] = [0x50, 0x4b, 0x03, 0x04];
pub const CENTRAL_DIRECTORY_FILE_SIGNATURE: [u8; 4] = [0x50, 0x4b, 0x01, 0x02];
pub const DATA_DESCRIPTOR_SIGNATURE: [u8; 4] = [0x08, 0x07, 0x4b, 0x50];
pub const END_CENTRAL_DIRECTORY_SIGNATURE: [u8; 4] = [0x50, 0x4b, 0x05, 0x06];

#[repr(transparent)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Os(pub u8);

impl Os {
    pub fn name(self) -> OsName {
        OsName::from_u8(self.0)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum OsName {
    Dos = 0,
    Amiga = 1,
    OpenVMS = 2,
    Unix = 3,
    VM = 4,
    AtariST = 5,
    OS2HPFS = 6,
    Macintosh = 7,
    ZSystem = 8,
    Cpm = 9,
    WindowsNtfs = 10,
    Mvs = 11,
    Vse = 12,
    AcornRisc = 13,
    Vfat = 14,
    AlternateMVS = 15,
    BeOS = 16,
    Tandem = 17,
    OS400 = 18,
    Darwin = 19,
    Unused,
}

impl OsName {
    pub fn from_u8(n: u8) -> OsName {
        match n {
            0 => OsName::Dos,
            1 => OsName::Amiga,
            2 => OsName::OpenVMS,
            3 => OsName::Unix,
            4 => OsName::VM,
            5 => OsName::AtariST,
            6 => OsName::OS2HPFS,
            7 => OsName::Macintosh,
            8 => OsName::ZSystem,
            9 => OsName::Cpm,
            10 => OsName::WindowsNtfs,
            11 => OsName::Mvs,
            12 => OsName::Vse,
            13 => OsName::AcornRisc,
            14 => OsName::Vfat,
            15 => OsName::AlternateMVS,
            16 => OsName::BeOS,
            17 => OsName::Tandem,
            18 => OsName::OS400,
            19 => OsName::Darwin,
            20..=255 => OsName::Unused,
        }
    }
}

#[repr(transparent)]
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct DateTimeModified(u32);

impl DateTimeModified {
    pub fn from_u32(b: u32) -> DateTimeModified {
        Self(b)
    }
}

#[repr(transparent)]
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct CompressionMethod(pub u16);

impl CompressionMethod {
    pub fn name(self) -> CompressionMethodName {
        CompressionMethodName::from_u16(self.0)
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum CompressionMethodName {
    None = 0,
    Shrink = 1,
    Factor1 = 2,
    Factor2 = 3,
    Factor3 = 4,
    Factor4 = 5,
    Implode = 6,
    Reserved = 7,
    Deflate = 8,
    EnhancedDeflate = 9,
    PKWareDclImplode = 10,
    Bzip2 = 12,
    Lzma = 14,
    IbmTerse = 18,
    IbmLZ77z = 19,
    PPMd = 98,
    Unknown,
}

impl CompressionMethodName {
    pub fn from_u16(n: u16) -> CompressionMethodName {
        match n {
            0 => CompressionMethodName::None,
            1 => CompressionMethodName::Shrink,
            2 => CompressionMethodName::Factor1,
            3 => CompressionMethodName::Factor2,
            4 => CompressionMethodName::Factor3,
            5 => CompressionMethodName::Factor4,
            6 => CompressionMethodName::Implode,
            7 | 11 | 13 | 15..=17 => CompressionMethodName::Reserved,
            8 => CompressionMethodName::Deflate,
            9 => CompressionMethodName::EnhancedDeflate,
            10 => CompressionMethodName::PKWareDclImplode,
            12 => CompressionMethodName::Bzip2,
            14 => CompressionMethodName::Lzma,
            18 => CompressionMethodName::IbmTerse,
            19 => CompressionMethodName::IbmLZ77z,
            98 => CompressionMethodName::PPMd,
            _ => CompressionMethodName::Unknown,
        }
    }
}

/// General purpose bit flags related to encoding
#[repr(transparent)]
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct ZipFlags(pub u16);

impl ZipFlags {
    pub const ENCRYPTED_FILE: u16 = 1 << 0;
    pub const COMPRESSION_OPTION1: u16 = 1 << 1;
    pub const COMPRESSION_OPTION2: u16 = 1 << 2;
    pub const DATA_DESCRIPTOR: u16 = 1 << 3;
    pub const ENHANCED_DEFLATION: u16 = 1 << 4;
    pub const COMPRESSED_PATCHED_DATA: u16 = 1 << 5;
    pub const STRONG_ENCRYPTION: u16 = 1 << 6;
    pub const LANGUAGE_ENCODING: u16 = 1 << 11;
    pub const RESERVED: u16 = 1 << 12;
    pub const MASK_HEADER_VALUES: u16 = 1 << 13;

    pub fn has_data_descriptor(&self) -> bool {
        (self.0 & Self::DATA_DESCRIPTOR) != 0
    }
}

#[repr(transparent)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ExternalAttributes(pub u32);

#[repr(transparent)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct InternalAttributes(pub u16);

impl InternalAttributes {
    pub const IS_ASCII: u16 = 1 << 0;
    pub const RESERVED: u16 = 1 << 1;
    pub const CONTROL_FIELD_RECORDS_PRECEDE_LOGICAL_RECORDS: u16 = 1 << 2;
}


src/main.rs
#![deny(missing_debug_implementations, unsafe_code)]

extern crate bitreader;
extern crate flate2;

use std::default::Default;
use std::fmt;
use std::fs::File;
use std::io;
use std::io::{BufRead, BufReader, Read};

use bitreader::BitReader;
use flate2::read::DeflateDecoder;

// TODO // u32::from_le_bytes(buffer).to_be_bytes()

/// ZIP file magic bytes
const LOCAL_FILE_SIGNATURE: [u8; 4] = [0x50, 0x4b, 0x03, 0x04];
const DATA_DESCRIPTOR_SIGNATURE: [u8; 4] = [0x08, 0x07, 0x4b, 0x50];
const CENTRAL_DIRECTORY_SIGNATURE: [u8; 4] = [0x50, 0x4b, 0x01, 0x02];
const END_CENTRAL_DIRECTORY_SIGNATURE: [u8; 4] = [0x50, 0x4b, 0x05, 0x06];

/// Read `n` bytes as [u8; n]
macro_rules! read_bytes_to_buffer {
    ($reader:expr, $bytes:literal) => {
        if let Some(mut buffer) = Some([0u8; $bytes]) {
            $reader.read_exact(&mut buffer)?;
            buffer
        } else {
            unreachable!()
        }
    };
}

/// Read 1 byte as a u8
macro_rules! read_u8 {
    ($reader:expr) => {
        if let Some(mut buffer) = Some([0u8]) {
            $reader.read_exact(&mut buffer)?;
            u8::from_le_bytes(buffer)
        } else {
            unreachable!()
        }
    };
}

/// Read 2 bytes as a u16
macro_rules! read_u16 {
    ($reader:expr) => {
        if let Some(mut buffer) = Some([0u8; 2]) {
            $reader.read_exact(&mut buffer)?;
            u16::from_le_bytes(buffer)
        } else {
            unreachable!()
        }
    };
}

/// Read 4 bytes as a u32
macro_rules! read_u32 {
    ($reader:expr) => {
        if let Some(mut buffer) = Some([0u8; 4]) {
            $reader.read_exact(&mut buffer)?;
            u32::from_le_bytes(buffer)
        } else {
            unreachable!()
        }
    };
}

#[derive(Debug)]
pub enum OS {
    DOS = 0,
    Amiga = 1,
    OpenVMS = 2,
    UNIX = 3,
    VM = 4,
    AtariST = 5,
    OS2HPFS = 6,
    Macintosh = 7,
    ZSystem = 8,
    CPM = 9,
    WindowsNTFS = 10,
    MVS = 11,
    VSE = 12,
    AcornRisc = 13,
    VFAT = 14,
    AlternateMVS = 15,
    BeOS = 16,
    Tandem = 17,
    OS400 = 18,
    Darwin = 19,
    Unused,
}

impl Default for OS {
    fn default() -> OS {
        OS::Unused
    }
}

impl OS {
    pub fn from_u8(n: u8) -> OS {
        match n {
            0 => OS::DOS,
            1 => OS::Amiga,
            2 => OS::OpenVMS,
            3 => OS::UNIX,
            4 => OS::VM,
            5 => OS::AtariST,
            6 => OS::OS2HPFS,
            7 => OS::Macintosh,
            8 => OS::ZSystem,
            9 => OS::CPM,
            10 => OS::WindowsNTFS,
            11 => OS::MVS,
            12 => OS::VSE,
            13 => OS::AcornRisc,
            14 => OS::VFAT,
            15 => OS::AlternateMVS,
            16 => OS::BeOS,
            17 => OS::Tandem,
            18 => OS::OS400,
            19 => OS::Darwin,
            20..=255 => OS::Unused,
        }
    }
}

#[derive(Debug, Default)]
pub struct InternalAttributes {
    is_ascii: bool,
    control_field_records_precede_logical_records: bool,
}

impl InternalAttributes {
    pub fn from_bytes(b: [u8; 2]) -> InternalAttributes {
        let mut bit_reader = BitReader::new(&b);
        let is_ascii = bit_reader.read_u8(1).unwrap() == 1u8;
        bit_reader.skip(1).unwrap();
        let control_field_records_precede_logical_records = bit_reader.read_u8(1).unwrap() == 1u8;
        InternalAttributes {
            is_ascii,
            control_field_records_precede_logical_records,
        }
    }
}

#[derive(Debug)]
pub enum ExternalAttributes {
    TODO,
}

impl Default for ExternalAttributes {
    fn default() -> ExternalAttributes {
        ExternalAttributes::TODO
    }
}

#[derive(Debug, Default)]
pub struct CentralDirectory {
    pub os: OS,
    pub metadata: ZippedFileMetadata,
    pub comment: Option<String>,
    pub disk_num_start: u16,
    pub internal_attributes: InternalAttributes,
    pub external_attributes: ExternalAttributes,
    pub zip_specification_version: u8,
    pub local_header_offset: u32,
}

#[derive(Debug, Default)]
pub struct EndCentralDirectory {
    pub disk_num: u16,
    pub disk_central_dir_num: u16,
    pub disk_entires: u16,
    pub total_entires: u16,
    pub central_dir_size: u32,
    pub central_dir_offset: u32,
    pub comment: Option<String>,
}
#[derive(Debug, Copy, Clone, Default)]
pub struct DateTimeModified {
    second: u8,
    minute: u8,
    hour: u8,
    day: u8,
    month: u8,
    // years since 1980
    year: u16,
}

impl fmt::Display for DateTimeModified {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{:0>4}-{:0>2}-{:0>2}T{:0>2}:{:0>2}:{:0>2}",
            self.year, self.month, self.day, self.hour, self.minute, self.second
        )
    }
}

impl DateTimeModified {
    pub fn from_bytes(mut b: [u8; 4]) -> DateTimeModified {
        b = u32::from_le_bytes(b).to_be_bytes();
        let mut bit_reader = BitReader::new(&b);
        let second = 2 * bit_reader.read_u8(5).unwrap();
        let minute = bit_reader.read_u8(6).unwrap();
        let hour = bit_reader.read_u8(5).unwrap();

        let day = bit_reader.read_u8(5).unwrap();
        let month = bit_reader.read_u8(4).unwrap();
        let year = 1980 + bit_reader.read_u16(7).unwrap();

        DateTimeModified {
            second,
            minute,
            hour,
            day,
            month,
            year,
        }
    }
}

#[derive(Debug, Copy, Clone)]
pub enum CompressionMethod {
    None = 0,
    Shrink = 1,
    Factor1 = 2,
    Factor2 = 3,
    Factor3 = 4,
    Factor4 = 5,
    Implode = 6,
    Reserved = 7,
    Deflate = 8,
    EnhancedDeflate = 9,
    PKWareDclImplode = 10,
    BZIP2 = 12,
    LZMA = 14,
    IbmTerse = 18,
    IbmLZ77z = 19,
    PPMd = 98,
}

impl CompressionMethod {
    pub fn from_i64(n: i64) -> CompressionMethod {
        match n {
            0 => CompressionMethod::None,
            1 => CompressionMethod::Shrink,
            2 => CompressionMethod::Factor1,
            3 => CompressionMethod::Factor2,
            4 => CompressionMethod::Factor3,
            5 => CompressionMethod::Factor4,
            6 => CompressionMethod::Implode,
            7 | 11 | 13 | 15..=17 => CompressionMethod::Reserved,
            8 => CompressionMethod::Deflate,
            9 => CompressionMethod::EnhancedDeflate,
            10 => CompressionMethod::PKWareDclImplode,
            12 => CompressionMethod::BZIP2,
            14 => CompressionMethod::LZMA,
            18 => CompressionMethod::IbmTerse,
            19 => CompressionMethod::IbmLZ77z,
            98 => CompressionMethod::PPMd,
            _ => unimplemented!(),
        }
    }

    pub fn from_u8(n: u8) -> CompressionMethod {
        CompressionMethod::from_i64(i64::from(n))
    }

    pub fn from_u16(n: u16) -> CompressionMethod {
        CompressionMethod::from_i64(i64::from(n))
    }
}

impl Default for CompressionMethod {
    fn default() -> CompressionMethod {
        CompressionMethod::None
    }
}

#[derive(Debug, Default)]
pub struct ZippedFileMetadata {
    /// PKZip version needed to unzip file
    pub version_needed: u16,
    pub compression_method: CompressionMethod,
    pub date_time_modified: DateTimeModified,
    /// Bit flags related to encoding
    pub flags: ZipFlags,
    /// File name
    pub name: String,
    pub compressed_size: u64,
    pub uncompressed_size: u64,
    pub crc: [u8; 4],
    /// Key value pairs of extra information
    pub extra_fields: Vec<u8>,
}

/// General purpose bit flags related to encoding
#[derive(Debug, Copy, Clone, Default)]
pub struct ZipFlags {
    is_encrypted: bool,
    /*
        (For Method 6 - Imploding)
        Bit 1: If the compression method used was type 6,
               Imploding, then this bit, if set, indicates
               an 8K sliding dictionary was used.  If clear,
               then a 4K sliding dictionary was used.

        Bit 2: If the compression method used was type 6,
               Imploding, then this bit, if set, indicates
               3 Shannon-Fano trees were used to encode the
               sliding dictionary output.  If clear, then 2
               Shannon-Fano trees were used.

        (For Methods 8 and 9 - Deflating)
        Bit 2  Bit 1
          0      0    Normal (-en) compression option was used.
          0      1    Maximum (-exx/-ex) compression option was used.
          1      0    Fast (-ef) compression option was used.
          1      1    Super Fast (-es) compression option was used.

        (For Method 14 - LZMA)
        Bit 1: If the compression method used was type 14,
               LZMA, then this bit, if set, indicates
               an end-of-stream (EOS) marker is used to
               mark the end of the compressed data stream.
               If clear, then an EOS marker is not present
               and the compressed data size must be known
               to extract.

        Note:  Bits 1 and 2 are undefined if the compression
               method is any other.

    // compression option
    // compression option
    */
    has_data_descriptor: bool,
    enhanced_deflation: bool,
    compressed_patched_data: bool,
    /// Strong encryption.  If this bit is set, you MUST
    /// set the version needed to extract value to at least
    /// 50 and you MUST also set bit 0.  If AES encryption
    /// is used, the version needed to extract value MUST
    /// be at least 51. See the section describing the Strong
    /// Encryption Specification for details.  Refer to the
    /// section in this document entitled "Incorporating PKWARE
    /// Proprietary Technology into Your Product" for more
    /// information.
    strong_encryption: bool,
    is_utf8: bool,
    /// Set when encrypting the Central Directory to indicate
    /// selected data values in the Local Header are masked to
    /// hide their actual values.  See the section describing
    /// the Strong Encryption Specification for details.  Refer
    /// to the section in this document entitled "Incorporating
    /// PKWARE Proprietary Technology into Your Product" for
    /// more information.
    mask_header_values: bool,
}

impl ZipFlags {
    pub fn from_bytes(b: [u8; 2]) -> ZipFlags {
        let mut bit_reader = BitReader::new(&b);
        let is_encrypted = bit_reader.read_u8(1).unwrap() == 1u8;
        // TODO: bits 1 and 2
        let _ = bit_reader.read_u8(1).unwrap() == 1u8;
        let _ = bit_reader.read_u8(1).unwrap() == 1u8;
        let has_data_descriptor = bit_reader.read_u8(1).unwrap() == 1u8;
        let enhanced_deflation = bit_reader.read_u8(1).unwrap() == 1u8;
        let compressed_patched_data = bit_reader.read_u8(1).unwrap() == 1u8;
        let strong_encryption = bit_reader.read_u8(1).unwrap() == 1u8;
        bit_reader.skip(3).unwrap();
        let is_utf8 = bit_reader.read_u8(1).unwrap() == 1u8;
        let mask_header_values = bit_reader.read_u8(1).unwrap() == 1u8;
        bit_reader.skip(2).unwrap();

        ZipFlags {
            is_encrypted,
            has_data_descriptor,
            enhanced_deflation,
            compressed_patched_data,
            strong_encryption,
            is_utf8,
            mask_header_values,
        }
    }
}

/// A single file within a ZIP archive
#[derive(Debug)]
pub struct ZippedFile {
    metadata: ZippedFileMetadata,
    data: Vec<u8>,
}

impl ZippedFile {
    /// Decompress bytes if necessary and return bytes
    pub fn data(self) -> Vec<u8> {
        match self.metadata.compression_method {
            CompressionMethod::None => self.data,
            CompressionMethod::Deflate => {
                let mut deflater = DeflateDecoder::new(&self.data[..]);
                let mut buffer = Vec::new();
                deflater.read_to_end(&mut buffer).unwrap();
                buffer
            }
            _ => unimplemented!("this compression method is currently unsupported"),
        }
    }
}

/// An entire ZIP archive
#[derive(Debug)]
pub struct ZippedArchive<R: Read + BufRead> {
    files: Vec<ZippedFile>,
    central_directory: CentralDirectory,
    end_central_directory: EndCentralDirectory,
    reader: R,
}

impl<R: Read + BufRead> ZippedArchive<R> {
    /// Construct ZippedArchive from and R such that R: Read + Bufread
    pub fn from_bufreader(r: R) -> Result<ZippedArchive<R>, io::Error> {
        ZippedArchive {
            files: Vec::new(),
            central_directory: Default::default(),
            end_central_directory: Default::default(),
            reader: r,
        }
        .unzip()
    }

    fn unzip(mut self) -> Result<ZippedArchive<R>, io::Error> {
        loop {
            let buf: [u8; 4] = read_bytes_to_buffer!(self.reader, 4);
            // Match on header using magic bytes
            match buf {
                LOCAL_FILE_SIGNATURE => self.read_file()?,
                CENTRAL_DIRECTORY_SIGNATURE => self.read_central_directory()?,
                END_CENTRAL_DIRECTORY_SIGNATURE => {
                    self.read_end_central_directory()?;
                    break;
                }
                _ => {
                    eprintln!("Unrecognized header: {:x?}", buf);
                    break;
                }
            };
        }

        Ok(self)
    }

    fn read_metadata(&mut self) -> Result<ZippedFileMetadata, io::Error> {
        let version_needed = read_u16!(self.reader);
        let bit_flags = ZipFlags::from_bytes(read_bytes_to_buffer!(self.reader, 2));
        let compression_method = CompressionMethod::from_u16(read_u16!(self.reader));
        let last_mod_date_time =
            DateTimeModified::from_bytes(read_bytes_to_buffer!(self.reader, 4));
        let crc: [u8; 4] = read_bytes_to_buffer!(self.reader, 4);
        let uncompressed_size = u64::from(read_u32!(self.reader));
        let compressed_size = u64::from(read_u32!(self.reader));
        let file_name_len = read_u16!(self.reader);
        let extra_field_len = read_u16!(self.reader);

        let mut file_name_buffer = vec![0u8; file_name_len as usize];
        self.reader.read_exact(&mut file_name_buffer)?;

        let mut extra_field_buffer = vec![0u8; extra_field_len as usize];
        self.reader.read_exact(&mut extra_field_buffer)?;

        let file_name = std::str::from_utf8(&file_name_buffer).unwrap().to_string();

        Ok(ZippedFileMetadata {
            version_needed,
            compression_method,
            date_time_modified: last_mod_date_time,
            flags: bit_flags,
            name: file_name,
            crc,
            compressed_size: compressed_size,
            uncompressed_size: uncompressed_size,
            extra_fields: Vec::from(extra_field_buffer),
        })
    }

    fn read_file(&mut self) -> Result<(), io::Error> {
        let mut metadata = self.read_metadata()?;

        if metadata.flags.has_data_descriptor {
            let optional_signature: [u8; 4] = read_bytes_to_buffer!(self.reader, 4);
            metadata.crc = if optional_signature == DATA_DESCRIPTOR_SIGNATURE {
                read_bytes_to_buffer!(self.reader, 4)
            } else {
                optional_signature
            };
            metadata.compressed_size = u64::from(read_u32!(self.reader));
            metadata.uncompressed_size = u64::from(read_u32!(self.reader));
        }

        let mut file_data_buffer = vec![0u8; metadata.uncompressed_size as usize];
        self.reader.read_exact(&mut file_data_buffer)?;

        dbg!(&metadata);

        self.files.push(ZippedFile {
            metadata,
            data: file_data_buffer,
        });
        Ok(())
    }

    fn read_central_directory(&mut self) -> Result<(), io::Error> {
        let os = OS::from_u8(read_u8!(self.reader));
        let zip_specification_version = read_u8!(self.reader);
        let version_needed = read_u16!(self.reader);
        let bit_flags = ZipFlags::from_bytes(read_bytes_to_buffer!(self.reader, 2));
        let compression_method = CompressionMethod::from_u16(read_u16!(self.reader));
        let date_time_modified =
            DateTimeModified::from_bytes(read_bytes_to_buffer!(self.reader, 4));
        let crc: [u8; 4] = read_bytes_to_buffer!(self.reader, 4);
        let uncompressed_size = u64::from(read_u32!(self.reader));
        let compressed_size = u64::from(read_u32!(self.reader));
        let file_name_len = read_u16!(self.reader);
        let extra_field_len = read_u16!(self.reader);
        let comment_len = read_u16!(self.reader);
        let disk_num_start = read_u16!(self.reader);
        let internal_attributes =
            InternalAttributes::from_bytes(read_bytes_to_buffer!(self.reader, 2));
        let external_attributes: [u8; 4] = read_bytes_to_buffer!(self.reader, 4);
        let local_header_offset = read_u32!(self.reader);

        let mut file_name_buffer = vec![0u8; file_name_len as usize];
        self.reader.read_exact(&mut file_name_buffer)?;

        let file_name = std::str::from_utf8(&file_name_buffer).unwrap().to_string();

        let mut extra_field_buffer = vec![0u8; extra_field_len as usize];
        self.reader.read_exact(&mut extra_field_buffer)?;

        let comment = if comment_len > 0 {
            let mut comment_buffer = vec![0u8; comment_len as usize];
            self.reader.read_exact(&mut comment_buffer)?;
            Some(std::str::from_utf8(&comment_buffer).unwrap().to_string())
        } else {
            None
        };

        let metadata = ZippedFileMetadata {
            version_needed,
            compression_method,
            date_time_modified,
            flags: bit_flags,
            name: file_name,
            crc,
            compressed_size: compressed_size,
            uncompressed_size: uncompressed_size,
            extra_fields: Vec::from(extra_field_buffer),
        };

        self.central_directory = CentralDirectory {
            os,
            comment,
            metadata,
            internal_attributes,
            external_attributes: ExternalAttributes::TODO,
            disk_num_start,
            zip_specification_version,
            local_header_offset,
        };
        Ok(())
    }

    fn read_end_central_directory(&mut self) -> Result<(), io::Error> {
        let disk_num = read_u16!(self.reader);
        let disk_central_dir_num = read_u16!(self.reader);
        let disk_entires = read_u16!(self.reader);
        let total_entires = read_u16!(self.reader);
        let central_dir_size = read_u32!(self.reader);
        let central_dir_offset = read_u32!(self.reader);
        let comment_len = read_u16!(self.reader);

        let comment = if comment_len > 0 {
            let mut comment_buffer = vec![0u8; comment_len as usize];
            self.reader.read_exact(&mut comment_buffer)?;
            Some(std::str::from_utf8(&comment_buffer).unwrap().to_string())
        } else {
            None
        };

        self.end_central_directory = EndCentralDirectory {
            disk_num,
            disk_central_dir_num,
            disk_entires,
            total_entires,
            central_dir_size,
            central_dir_offset,
            comment,
        };

        Ok(())
    }

    /// Consume self and return list of files
    pub fn files(self) -> Vec<ZippedFile> {
        self.files
    }
}

impl ZippedArchive<BufReader<File>> {
    pub fn from_path<P: AsRef<std::path::Path>>(
        p: P,
    ) -> Result<ZippedArchive<BufReader<File>>, io::Error> {
        let buffer = BufReader::new(File::open(p).unwrap());
        ZippedArchive::from_bufreader(buffer)
    }
}

#[cfg(test)]
mod test {}

fn main() -> io::Result<()> {
    let zip = ZippedArchive::from_path(FILE_PATH).unzip()?;

    // dbg!(bit_reader.read_u8(2).unwrap());
    Ok(())
}
